### PR TITLE
Feature Support API (Part 2)

### PR DESCRIPTION
**Pull request by Sanithovski**
This is the second part of the new feature support (class) APIs. The branch is constantly synced with the first part so all modifications suggested in the previous PR review is pushed to this branch as well. The new code is also updated to stay consistent with part 1. Once the part 1 PR is merged you'll be able to see the code changes more clearly, but you can take a look right now at the new features and check routines. I will forward the open questions to this PR.

I'll make inline comments about code snippets for potential discussion, but you're always welcome to provide feedback anywhere in the code.

### CODE REVIEW

File: include/directx/d3dx12.h [hunk 1]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions10, VariableRateShadingSumCombinerSupported);
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions10, MeshShaderPerPrimitiveShadingRateSupported);
+
+    // 40: Options11
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions11, AtomicInt64OnDescriptorHeapResourceSupported);
+
+private: // Private structs to simplify querying procedures
+    
+    // A helper struct that help store protected resource session types
+    struct ProtectedResourceSessionTypesLocal : D3D12_FEATURE_DATA_PROTECTED_RESOURCE_SESSION_TYPES
```
Comment (Sanithovski) [inline include/directx/d3dx12.h]: I created a custom struct hold the data for the resource session types check. See below for more information.

File: include/directx/d3dx12.h [hunk 2]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+
+            if (result == DXGI_ERROR_UNSUPPORTED) { // Indicates that none are supported
+                result = S_OK;
+            }
+        }
+        return result;
+    }
+
+    // Helper function to initialize local protected resource session types structs
+    HRESULT QueryProtectedResourceSessionTypes(UINT NodeIndex, UINT Count)
```
Comment (Sanithovski) [inline include/directx/d3dx12.h]: This is perhaps the only special check routine for part 2. 

1. It is one of the features that requires one feature data struct for each graphics node.
2. One of its in parameters depends on the results of another feature check, ProtectedResourceSessionTypeCount. Thus this check must be completed after getting the count in Init().
2. CheckFeatureSupport will take in a pointer (pTypes) and a size (Count) and store the output as a (C-style) array into the given pointer. The size is decidable from the ProtectedResourceSessionTypeCount before the API call, but it's probably not known at compile time. 

To avoid manually managing dynamic memory allocation, I've decided to extend the original D3D12_FEATURE_SUPPORT_DATA_PROTECTED_RESOURCE_SESSION_TYPES and add a vector to it. Upon initialization I will resize the vector to fit the type count and pass the underlying array's pointer (through vector::data()) to pTypes. When CheckFeatureSupport writes the output data to pTypes, it's essentially writing data to the vector. 

Finally, in the new checking API I simply returns the vector.

File: test/feature_check_test.cpp [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+    // 30: Options6
+    BOOL AdditionalShadingRatesSupported = features->AdditionalShadingRatesSupported();
+    BOOL PerPrimitiveShadingRateSupportedWithViewportIndexing = features->PerPrimitiveShadingRateSupportedWithViewportIndexing();
+    D3D12_VARIABLE_SHADING_RATE_TIER VariableShadingRateTier = features->VariableShadingRateTier();
+    BOOL ShadingRateImageTileSize = features->ShadingRateImageTileSize();
+    BOOL BackgroundProcessingSupported = features->BackgroundProcessingSupported();
+
+    // 31: Query Metacommand
+    // TODO: Don't know how to test
```
Comment (Sanithovski) [inline test/feature_check_test.cpp]: Query metacommand requires a GUID as an input parameter and I don't know how to use it :( Does someone know how to test this feature check?

File: include/directx/d3dx12.h [hunk 4]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+        }
+
+        if (INITIALIZE_FAILED(D3D12_FEATURE_EXISTING_HEAPS, m_dExistingHeaps)) {
+            m_dExistingHeaps.Supported = false;
+        }
+
+        if (INITIALIZE_FAILED(D3D12_FEATURE_D3D12_OPTIONS4, m_dOptions4)) {
+            m_dOptions4.MSAA64KBAlignedTextureSupported = false;
+            m_dOptions4.Native16BitShaderOpsSupported = false;
+            m_dOptions4.SharedResourceCompatibilityTier = (D3D12_SHARED_RESOURCE_COMPATIBILITY_TIER)0; // TODO: 0 is occupied. See if needs another value
```
Comment (Sanithovski) [inline include/directx/d3dx12.h]: 0 is sometimes a valid entry for some feature support data, particularly *_TIER enums that have a TIER_0 entry. Is this the same as NOT_SUPPORTED? If not, what default value would you suggest if the initialization fails?

File: include/directx/d3d12.h [hunk 5]
```diff
@@ -2300,7 +2302,8 @@ enum D3D_SHADER_MODEL
         D3D_SHADER_MODEL_6_4	= 0x64,
         D3D_SHADER_MODEL_6_5	= 0x65,
         D3D_SHADER_MODEL_6_6	= 0x66,
-        D3D_SHADER_MODEL_6_7	= 0x67
+        D3D_SHADER_MODEL_6_7	= 0x67,
+        D3D_HIGHEST_SHADER_MODEL = D3D_SHADER_MODEL_6_7 // TODO: Pull from latest d3d12.h
```
Comment (clandrew) [inline include/directx/d3d12.h]: Is there something more you're planning on doing here? Seems like there is just a manual step of keeping this updated

File: include/directx/d3dx12.h [hunk 6]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+    HRESULT Init(ID3D12Device* pDevice)
+    {
+        if (!pDevice) {
+            m_hStatus = E_INVALIDARG;
+            return m_hStatus;
+        }
+
+        m_pDevice = pDevice;
+
+
```
Comment (clandrew) [inline include/directx/d3dx12.h]: Small nit: extra space

File: include/directx/d3dx12.h (hunk 4 above)
Comment (clandrew) [reply include/directx/d3dx12.h]: Some features don't have a NOT_SUPPORTED, because there is always some level of support in D3D12. E.g., for shared resources, the minimum is TIER_0. A valid D3D12 implementation won't indiscriminately fail all attempts to use shared resources.

File: include/directx/d3dx12.h [hunk 7]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+        D3D_ROOT_SIGNATURE_VERSION allRootSignatureVersions[] = {
+            D3D_ROOT_SIGNATURE_VERSION_1_1,
+            D3D_ROOT_SIGNATURE_VERSION_1_0,
+            D3D_ROOT_SIGNATURE_VERSION_1,
+        };
+
+        UINT numRootSignatureVersions = sizeof(allRootSignatureVersions) / sizeof(D3D_ROOT_SIGNATURE_VERSION);
+
+
+        // ROOT_SIGNATURE_VERSION_1_X
```
Comment (clandrew) [inline include/directx/d3dx12.h]: Why is this here?

File: include/directx/d3dx12.h [hunk 8]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+    {}
+
+    CD3DX12FeatureSupport(ID3D12Device* pDevice)
+    {
+        Init(pDevice);
+    }
+
+    HRESULT Init(ID3D12Device* pDevice)
+    {
+        if (!pDevice) {
```
Comment (clandrew) [inline include/directx/d3dx12.h]: Very minor style thing: elsewhere in this file and in general, we use
if (cond)
{
}

not
if (cond) {
}

File: include/directx/d3dx12.h [hunk 9]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+    D3D12_FEATURE_DATA_COMMAND_QUEUE_PRIORITY m_dCommandQueuePriority;
+    D3D12_FEATURE_DATA_D3D12_OPTIONS3 m_dOptions3;
+    D3D12_FEATURE_DATA_EXISTING_HEAPS m_dExistingHeaps;
+    D3D12_FEATURE_DATA_D3D12_OPTIONS4 m_dOptions4;
+    std::vector<D3D12_FEATURE_DATA_SERIALIZATION> m_dSerialization; // Cat2 NodeIndex
+    D3D12_FEATURE_DATA_CROSS_NODE m_dCrossNode;
+    D3D12_FEATURE_DATA_D3D12_OPTIONS5 m_dOptions5;
+    D3D12_FEATURE_DATA_DISPLAYABLE m_dDisplayable;
+    D3D12_FEATURE_DATA_D3D12_OPTIONS6 m_dOptions6;
+    // D3D12_FEATURE_DATA_QUERY_META_COMMAND m_dQueryMetaCommand; // Cat3
```
Comment (clandrew) [inline include/directx/d3dx12.h]: Is this supposed to be here?

File: test/feature_check_test.cpp [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+     auto features_requested = CD3DX12FeatureSupport::Create(device);
+     if (auto failure = std::get_if<HRESULT>(&features_requested)) {
+         return -1;
+     }
+
+     features = std::get_if<CD3DX12FeatureSupport>(&features_requested);
+     if (!features) {
+         return -1;
+     }
+    // CD3DX12FeatureSupport* features = new CD3DX12FeatureSupport(device);
```
Comment (clandrew) [inline test/feature_check_test.cpp]: unneeded commented-out line?

File: test/feature_check_test.cpp [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+
+        IDXCoreAdapterList *list = nullptr;
+        if (FAILED(factory->CreateAdapterList(1, &DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE, &list)))
+            return -1;
+        
+        if (FAILED(list->GetAdapter(0, &adapter)))
+            return -1;
+    }
+
+    if (FAILED(D3D12CreateDevice(adapter, D3D_FEATURE_LEVEL_11_0, IID_PPV_ARGS(&device)))) {
```
Comment (clandrew) [inline test/feature_check_test.cpp]: If you just want the default adapter, is there a reason you don't just pass a null adapter? Maybe this is using a non default adapter in some manner I'm not seeing

File: test/feature_check_test.cpp (hunk 3 above)
Comment (clandrew) [reply test/feature_check_test.cpp]: Responded externally

**Review by clandrew — CHANGES_REQUESTED** [review]
Thanks for sending this out! I made some comments

File: include/directx/d3dx12.h [hunk 12]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+}
+
+// Macro to initialize a member feature data and report on failure
+// Does not stop the initialization process by itself
+#define INITIALIZE_FAILED(FEATURE, MEMBER) FAILED(m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))
+
+class CD3DX12FeatureSupport
+{
+public:
+    CD3DX12FeatureSupport()
```
Comment (amarpMSFT) [inline include/directx/d3dx12.h]: Continuation of my comment from preview PR:   I propose putting a raw declaration of all the member functions at the top of the class here.  So users can quickly see all the things they can do.  They can scroll down to see implementations if they want (and you can keep the macros there).

File: test/feature_check_test.cpp (hunk 3 above)
Comment (shawnhar) [reply test/feature_check_test.cpp]: Metacommands are kindof a unique feature: used for very specialist purpose by only a few developers.  I think maybe we should just leave them out of this helper API entirely.

File: include/directx/d3d12.h [hunk 13]
```diff
@@ -2257,7 +2257,9 @@ enum D3D_ROOT_SIGNATURE_VERSION
     {
         D3D_ROOT_SIGNATURE_VERSION_1	= 0x1,
         D3D_ROOT_SIGNATURE_VERSION_1_0	= 0x1,
-        D3D_ROOT_SIGNATURE_VERSION_1_1	= 0x2
```
Comment (shawnhar) [inline include/directx/d3d12.h]: This second part would be easier to review if you stage it as a pull request into the first branch, so it would show only the changes made since that first review as diffs.

File: include/directx/d3dx12.h [hunk 14]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+    , m_hStatus(E_INVALIDARG)
+    {}
+
+    CD3DX12FeatureSupport(ID3D12Device* pDevice)
+    {
+        Init(pDevice);
+    }
+
+    HRESULT Init(ID3D12Device* pDevice)
+    {
```
Comment (Sanithovski) [inline include/directx/d3dx12.h:4296]: Continuation from the last PR: Need to figure out how to report the errors to the user in a more intuitive manner.

Right now the status of initialization is stored in m_hStatus. User can use GetStatus() to access this underlying variable, but they must do so explicitly. The factory method returns a variant, so if it will return the error code only if the initialization failed, but that only works on C++ 17 compilers and if the user chooses to turn it on. Also the current factory method design will have different function signatures when the std::variant usage is turned on or off, which is not friendly for the users.

Comment (Sanithovski) [reply include/directx/d3dx12.h:4296]: Another question is whether we should disable the regular constructor and only allows users to create a CD3D12XFeatureSupport object through the factory method? The regular constructor cannot return anything but the constructed object. It can throw an exception, but not all programmers would want to use exception handling on a constructor.

File: include/directx/d3dx12.h (hunk 4 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: The I guess it is safe to use TIER_0 as a fallback value when the initialization fails, as long as we're running with a valid D3D12 implementation?

File: include/directx/d3d12.h (hunk 5 above)
Comment (Sanithovski) [reply include/directx/d3d12.h]: Probably no. And actually I don't need it any more since I've changed the process of iterating through all the shader models during initialization. I guess I will remove the extra entry after all. Thanks for the notice!

File: include/directx/d3dx12.h (hunk 7 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Oh. It's a leftover comment from a previous comment. I'll remove it in the next commit.

File: include/directx/d3dx12.h (hunk 12 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Agreed. Will do that in the next commit!

File: include/directx/d3dx12.h (hunk 9 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: It's just reminding me the order of the D3D12 features. I will remove this comment.

File: test/feature_check_test.cpp (hunk 11 above)
Comment (Sanithovski) [reply test/feature_check_test.cpp]: Frankly speaking I don't know the steps to set up a Device object, so I borrowed the code from the `test.cpp` in the same folder. There's nothing special I want from the adapter, and I will remove it if can be left blank.

File: include/directx/d3dx12.h (hunk 8 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: I see. This is also the way how the rest of the code is written. I will fix it before I finalize this PR since fixing this will bring a lot of code changes and make the "Code Changes" view hard to navigate.

File: test/feature_check_test.cpp (hunk 10 above)
Comment (Sanithovski) [reply test/feature_check_test.cpp]: This is actually another way to initialize a new FeatureSupport object -- through the constructor. I may remove this constructor in the future, but I guess I'll keep it for now just as a reminder.

Comment (shawnhar) [reply test/feature_check_test.cpp]: It'd be a good practice to comment that intention.  Something like:

// TODO either test this constructor option or remove if I decide not to keep the constructor

That way the intent will be clear to anyone else reading this code in the future.  Otherwise, if you forget to update this for some reason, it could be a confusing thing for somebody to figure out years later.

File: include/directx/d3dx12.h (hunk 14 above)
Comment (shawnhar) [reply include/directx/d3dx12.h:4296]: I would support getting rid of the non-default constructor, which would then also allow making GetStatus private.  This would leave two construction options:

1. The factory method (returning variant)
2. Default constructor followed by Init() that returns HRESULT

Both of these make error handling obvious and robust, without requiring use of exceptions.

File: include/directx/d3dx12.h [hunk 15]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+
+    // 2: Feature Levels
+    // Simply returns the highest supported feature level
+    D3D_FEATURE_LEVEL HighestFeatureLevel() const
+    {
+        return m_eMaxFeatureLevel;
+    }
+
+    // 3: Feature Format Support
+    HRESULT FormatSupport(DXGI_FORMAT Format, D3D12_FORMAT_SUPPORT1& Support1, D3D12_FORMAT_SUPPORT2& Support2)
```
Comment (Sanithovski) [inline include/directx/d3dx12.h]: Continuation from the previous PR: FormatSupport, MultisampleQualityLevels, FormatInfo all use DXGI_FORMAT. Is it possible to combine them into one feature check?

Another concern is that there are two output parameters from this function (Support1 and Support2). Is that too complicated for the users?

File: include/directx/d3dx12.h (hunk 4 above)
Comment (shawnhar) [reply include/directx/d3dx12.h]: Correct.  If there is no NOT_SUPPORTED option, the default should be whatever other enum value comes first (typically TIER_0).

File: include/directx/d3dx12.h [hunk 16]
```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+
+    // Feature support data structs
+    D3D12_FEATURE_DATA_D3D12_OPTIONS m_dOptions;
+    D3D_FEATURE_LEVEL m_eMaxFeatureLevel;
+    D3D12_FEATURE_DATA_GPU_VIRTUAL_ADDRESS_SUPPORT m_dGPUVASupport;
+    D3D12_FEATURE_DATA_SHADER_MODEL m_dShaderModel;
+    D3D12_FEATURE_DATA_D3D12_OPTIONS1 m_dOptions1;
+    std::vector<D3D12_FEATURE_DATA_PROTECTED_RESOURCE_SESSION_SUPPORT> m_dProtectedResourceSessionSupport;
+    D3D12_FEATURE_DATA_ROOT_SIGNATURE m_dRootSignature;
+    std::vector<D3D12_FEATURE_DATA_ARCHITECTURE1> m_dArchitecture1;
```
Comment (Sanithovski) [inline include/directx/d3dx12.h]: Continuation from last PR: std::vector is dynamic allocation which can cause memory failure during actual runs. Is it possible to replace these vectors (whose size is dependent on the number of graphics node in the machine) with statically allocated arrays, such as std::array?

**Comment by Sanithovski** [comment]
I've merged the first PR. Interestingly, GitHub didn't update the code diff when the target branch changed.

File: test/feature_check_test.cpp (hunk 11 above)
Comment (Sanithovski) [reply test/feature_check_test.cpp]: OK I verified that the adapter is not required. I'll remove all the unnecessary preparation code.

File: include/directx/d3dx12.h (hunk 12 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: It seems C++ doesn't allow me to declare the function then define the same function inside the class declaration scope. I'll have to do the definition outside the scope (moving the implementation out and scope them with CD3DX12FeatureSupport). This is a little bit more complicated than I expect. Since I'm running a little short on time, I'll perform this update when I finalize the PR. Sorry!

File: test/feature_check_test.cpp (hunk 11 above)
Comment (Sanithovski) [reply test/feature_check_test.cpp]: Resolved

File: include/directx/d3dx12.h (hunk 9 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Resolved

File: include/directx/d3dx12.h (hunk 7 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Resolved.

File: include/directx/d3dx12.h (hunk 6 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Resolved.

File: include/directx/d3d12.h (hunk 5 above)
Comment (Sanithovski) [reply include/directx/d3d12.h]: Resolved.

File: include/directx/d3dx12.h [hunk 17]
```diff
@@ -4037,6 +4037,660 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+#define FEATURE_SUPPORT_GET_NODE_INDEXED_NAME(RETTYPE,FEATURE,OPTION,NAME) \
+RETTYPE NAME(UINT NodeIndex=0) const \
+{\
+    return FEATURE[NodeIndex].OPTION; \
+}
+
+// Macro to initialize a member feature data struct corresponding to the specified feature
+// Will stop the initialization process and return an error code upon failure
+#define INITIALIZE_MEMBER_DATA_CHECKED(FEATURE, MEMBER) \
+if (FAILED(m_hStatus = m_pDevice->CheckFeatureSupport(FEATURE, &MEMBER, sizeof(MEMBER)))) { \
```
Comment (alexgong14) [inline include/directx/d3dx12.h]: nit: it looks better if you put { \ in a new line

File: include/directx/d3dx12.h (hunk 4 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: I'm using TIER_0 as default values for these caps. They'll be up in the next commit. Let me know if this can be an issue.

File: include/directx/d3dx12.h [hunk 18]
```diff
@@ -4037,6 +4037,660 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+    FEATURE_SUPPORT_GET_NAME(BOOL, m_dCrossNode, AtomicShaderInstructions, CrossNodeAtomicShaderInstructions);
+    
+    // 27: D3D12 Options5
+    FEATURE_SUPPORT_GET(BOOL, m_dOptions5, SRVOnlyTiledResourceTier3);
+    FEATURE_SUPPORT_GET(D3D12_RENDER_PASS_TIER, m_dOptions5, RenderPassesTier);
+    FEATURE_SUPPORT_GET(D3D12_RAYTRACING_TIER, m_dOptions5, RaytracingTier);
+
+    // 28: Displayable
+    FEATURE_SUPPORT_GET(BOOL, m_dDisplayable, DisplayableTexture);
+    FEATURE_SUPPORT_GET_NAME(D3D12_SHARED_RESOURCE_COMPATIBILITY_TIER, m_dDisplayable, SharedResourceCompatibilityTier, DisplayableSharedResourceCompatibilityTier);
```
Comment (Sanithovski) [inline include/directx/d3dx12.h]: There is another `SharedResourceCompatibilityTier` cap in D3D12Options4. I checked the code in Device.cpp and found that while the code to check for these two caps are slightly different, the `SharedResourceCompatibilityTier` from `Displayable` is actuallyreading from `m_ValidationInfo.m_Options4.SharedResourceCompatibilityTier`. Does it suggest the two are essentially the same?

**Comment by Sanithovski** [comment]
Updates in the latest commit:
- Fixed an issue with caps that exist in more than one feature. See the following comments for details.
- Renamed loop variables to be more descriptive.
- Use TIER_0 as default values for applicable caps.
- Minor format fix.

File: include/directx/d3dx12.h [hunk 19]
```diff
@@ -4037,6 +4037,683 @@ class CD3DX12_NODE_MASK_SUBOBJECT
...
+    // Special procedure for handling caps that is also part of other features
+    D3D12_CROSS_NODE_SHARING_TIER CrossNodeSharingTier() const
+    {
+        if (m_dCrossNode.SharingTier > D3D12_CROSS_NODE_SHARING_TIER_NOT_SUPPORTED)
+        {
+            return m_dCrossNode.SharingTier;
+        }
+        else
+        {
+            return m_dOptions.CrossNodeSharingTier;
```
Comment (Sanithovski) [inline include/directx/d3dx12.h]: Some caps such as `CrossNodeSharingTier` appears in more than one feature data struct. While one of these structs are not recognized by the runtime, the other feature may still be available and provides support for the cap. Fortunately in most cases the two caps get their information from the same variable/function. I added a check for these features so they will check for both caps and return the one that's properly initialized.

File: test/feature_check_test.cpp (hunk 3 above)
Comment (Sanithovski) [reply test/feature_check_test.cpp]: The new API basically makes a forwarding call to the original API, so there's not much to test to ensure its correctness besides making sure it passes the params correctly. I'm going to skip it for now but I will write some basic test cases in the test suite.

File: include/directx/d3dx12.h (hunk 4 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Resolved.

File: include/directx/d3dx12.h (hunk 17 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Resolved.

**Comment by Sanithovski** [comment]
I've decided to keep only one initialization method for the class:
1. User create an empty object using the default constructor.
2. User calls the `Init()` function to initialize the object. User can read the HRESULT returned by `Init()` to learn if the construction succeeds or fails.

While this is not the most elegant solution, I believe it is the most straightforward way to resolve the issue and requires minimal changes to the current code. Although we can't enforce the users to check the returned HRESULT, just like other factory methods in D3D. 

* A classical constructor cannot return an HRESULT as a return value. Users may forget to check whether the initialization succeeds. While we can use `throw ... catch`, it's not typical for a constructor to throw exceptions and not favored by everyone. 
* The `Create()` method which returns std::variant or std::optional can provide additional benefits, but they require C++17 and not all users are familiar with them. Without the two C++17 libraries the method does nothing more than a constructor.
* We can create a classical factory method like `HRESULT CreateFeatureSupport(ID3D12Device* pDevice, CD3DX12FeatureSupport* pFeatures)`, but essentially this is a transformation of the "Construct -> Init" process with a more complicated signature.

File: include/directx/d3dx12.h (hunk 14 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h:4296]: I've decided to keep only the default constructor and the Init() method. See the new comment below.

File: test/feature_check_test.cpp (hunk 3 above)
Comment (clandrew) [reply test/feature_check_test.cpp]: Makes sense!

File: include/directx/d3dx12.h (hunk 4 above)
Comment (clandrew) [reply include/directx/d3dx12.h]: Yeah, sounds good.

File: include/directx/d3dx12.h (hunk 8 above)
Comment (clandrew) [reply include/directx/d3dx12.h]: That's very considerate of you to keep the diff clean. Sure thing.

File: include/directx/d3dx12.h (hunk 18 above)
Comment (clandrew) [reply include/directx/d3dx12.h]: That's a good question. From what I can tell, the Displayable field used to be experimental, and capable of returning values that Option4 would not. However, now the Displayable field isn't experimental any more, so it would appear to be a redundant cap. @alexgong14 might know more about it.

Aside: in the API there's caps that are definitely redundant, and I have trouble imagining that changing: options.MaxGPUVirtualAddressBitsPerResource and GpuVirtualAddressSupport.MaxGPUVirtualAddressBitsPerResource.

Comment (alexgong14) [reply include/directx/d3dx12.h]: The short answer is Yes.

Displayable surface is a D3D11 feature and we haven't yet allowed the use of it in D3D12. If you call D3D12 CheckFeatureSupport, the field DisplayableTexture will always be False. This indicates it is not supported by D3D12. For SharedResourceTier, both D3D11 Displayable and D3D12 Displayable share it with other D3D11/D3D12Options, the Output below is used by both D3D11_FEATURE_D3D11_OPTIONS5 and D3D11_FEATURE_DISPLAYABLE:

D3D11_SHARED_RESOURCE_TIER Output = D3D11_SHARED_RESOURCE_TIER_0;
            if (!IsExplicitReferenceDriver())
            {
                Output = D3D11_SHARED_RESOURCE_TIER
                    ( D3D11_SHARED_RESOURCE_TIER_0 + CD3D11FormatHelper::MediaSharedResourceTier
                        ( GetExtendedFormatFeatures()
                        , true ) // Feature_D3D1XDisplayable
                    );
            }
            
In my opinion, the reason Displayable includes SharedResourceTier is just for convenience, because Displayable requires current WDDM version to be no less than a certain WDDM Version, for D3D11 it is WDDM 2.7. So when people try to use Displayable but they get DisplayableTexture = false, they can also look at the SharedResourceTier to see if Displayable is not supported because their device does not meet the minimum requirement to support Displayable.

File: test/feature_check_test.cpp (hunk 10 above)
Comment (Sanithovski) [reply test/feature_check_test.cpp]: I've removed this constructor so this comment is not necessary. I will remove it in the next commit.

File: include/directx/d3dx12.h (hunk 18 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Got it. Thanks for the explanation! If these two caps are the same thing, then I no longer need the `DisplayableSharedResourceTier` feature any more. I will remove this redundant API in the next commit.

**Comment by Sanithovski** [comment]
I've made a new commit that addressed the discussed issues. Though the code looks quite different, I've only made some changes in code structure and formatting with little changes on the actual contents:
- Kept only the default constructor and the Init() function. Removed Create() and the customized constructor.
- Removed a duplicate SharedResourceCompatibilityTier getter function from the Displayable feature.
- Moved opening curly braces to new lines.
- Separate function declaration from definition. Moved definition out of the class.
- Replaced some macros with actual code.
- Updated basic test files.

File: include/directx/d3dx12.h (hunk 8 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Resolved.

File: test/feature_check_test.cpp (hunk 10 above)
Comment (Sanithovski) [reply test/feature_check_test.cpp]: Resolved

File: include/directx/d3dx12.h (hunk 12 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Resolved. I've separated declarations from definitions and moved definitions to the bottom.

File: include/directx/d3dx12.h (hunk 18 above)
Comment (Sanithovski) [reply include/directx/d3dx12.h]: Resolved. The duplicate getter is removed.

**Review by shawnhar — APPROVED** [review]
